h/services/search_index/_queue.py

```diff
@@ -0,0 +1,146 @@
+import datetime
+import logging
+
+from dateutil.parser import isoparse
+
+from h.models import Annotation, Job
+
+logger = logging.getLogger(__name__)
+
+
+class Queue:
+    """A job queue for synchronizing annotations from Postgres to Elastic."""
+
+    def __init__(self, db, es, batch_indexer, limit):
+        self._db = db
+        self._es = es
+        self._batch_indexer = batch_indexer
+        self._limit = limit
+
+    def add(self, annotation_id, tag, scheduled_at=None):
+        """Queue an annotation to be synced to Elasticsearch."""
+        self.add_all([annotation_id], tag, scheduled_at)
+
+    def add_all(self, annotation_ids, tag, scheduled_at=None):
```

Comment:
Would `scheduled_at` be better as a delay (from now) instead?

Reply:
Hmm.

`scheduled_at` needs to be a datetime not a duration in the DB because we filter by it--if it was a duration on top of `enqueued_at` in the DB then Postgres would have to calculate `enqueued_at` + `scheduled_at` for every row in the DB every time we want to retrieve jobs.

So we'd have to implement this by having the Python service accept a duration and transform that into a datetime for the DB. Something like:

```python
def add_all(..., scheduled_at=None):
    scheduled_at = utcnow() + scheduled_at
    ...
```

I'm calculating the date from `utcnow()` not from `enqueued_at`. The Python code can't do `scheduled_at = enqueued_at + scheduled_at` because `enqueued_at` is not computed by Postgres until it commits the rows (`server_default`), it's not available to Python until after the rows have been committed. Except maybe if you flush the DB first but that's gonna be awkward and inefficient especially when adding a lot of rows. Calculating from `utcnow()` should be fine though I'd think.

Reply:
:+1: Done

Reply:
Gah, I really should have said, if we're going to change the type the argument name should change too.

In the DB it's `scheduled_at` but here it's probably `scheduled_in` until it's added to now.

Reply:
Done: https://github.com/hypothesis/h/pull/6265/commits/3f9996cd5f153e98e2cb6f609533c3ad626750af I just couldn't think of another name for it, but `schedule_in` works